python/lsst/daf/persistence/butler.py

```diff
@@ -229,35 +437,49 @@ def getKeys(self, datasetType=None, level=None):
         @returns (dict) valid data id keys; values are corresponding types.
         """
         datasetType = self._resolveDatasetTypeAlias(datasetType)
-        return self.repository.getKeys(datasetType, level)
+
+        keys = None
+        for repoData in self._repos.inputs():
+            if not tag or len(tag.intersection(repoData.tags)) > 0:
+                keys = repoData.repo.getKeys(datasetType, level)
+                # An empty dict is a valid "found" condition for keys. The only value for keys that should
+                # cause the search to continue is None
+                if keys is not None:
+                    break
+        return keys
 
 
     def queryMetadata(self, datasetType, format=None, dataId={}, **rest):
         """Returns the valid values for one or more keys when given a partial
         input collection data id.
 
-        @param datasetType (str)    the type of dataset to inquire about.
-        @param key (str)            a key giving the level of granularity of the inquiry.
-        @param format (str, tuple)  an optional key or tuple of keys to be returned.
-        @param dataId (dict)        the partial data id.
-        @param **rest               keyword arguments for the partial data id.
+        @param datasetType (str) the type of dataset to inquire about.
+        @param key (str) a key giving the level of granularity of the inquiry.
+        @param format (str, tuple) an optional key or tuple of keys to be returned.
+        @param dataId (DataId, dict) the partial data id.
+        @param **rest keyword arguments for the partial data id.
         @returns (list) a list of valid values or tuples of valid values as
         specified by the format (defaulting to the same as the key) at the
         key's level of granularity.
         """
 
         datasetType = self._resolveDatasetTypeAlias(datasetType)
-        dataId = copy.copy(dataId)
+        dataId = DataId(dataId)
         dataId.update(**rest)
```

Comment:
The most natural way for users to specify a tag in a butler call (IMO) is something like:
```
calexp = butler.get("calexp", visit=236, ccd=44, tag="cfht")
```
If the intent is to force them to instead do
```
calexp = butler.get("calexp", DataId(visit=236, ccd=44, tag="cfht"))
```
that's not terrible, but it needs to be well-documented to avoid causing confusion.